pkg/plugin/trans_test.go

```diff
@@ -0,0 +1,521 @@
+			podAfter: pod{
+				reserved: 3,
+				buffer:   0,
+				pressure: 2,
+			},
+			nodeAfter: node{
+				reserved: 10,
+				buffer:   0,
+				pressure: 2,
+			},
+		},
+		{
+			name:   "UpscaleRejectedAtLimitsWithUnevenFactor",
+			podMin: 2,
+			podMax: 8,
+			podBefore: pod{
+				reserved: 4,
+				buffer:   0,
+				pressure: 0,
+			},
+			nodeTotal: 10,
+			nodeBefore: node{
+				reserved: 7,
+				buffer:   0,
+				pressure: 0,
+			},
+
+			requested:  8,
+			lastPermit: lo.ToPtr[uint](4),
+			factor:     2,
+
+			// FIXME: This is actually incorrect / not working. The request should be rejected with
+			// only 6 reserved, because otherwise we end up over the total (as seen below).
+			verdict: "",
+			podAfter: pod{
+				reserved: 8,
+				buffer:   0,
+				pressure: 0,
+			},
+			nodeAfter: node{
+				reserved: 11,
+				buffer:   0,
+				pressure: 0,
+			},
+		},
+		{
+			name:   "UpscaleRejectedAboveLimits",
+			podMin: 1,
+			podMax: 4,
+			podBefore: pod{
+				reserved: 2,
+				buffer:   0,
+				pressure: 0,
+			},
+			nodeTotal: 10,
+			nodeBefore: node{
+				reserved: 12,
+				buffer:   0,
+				pressure: 0,
+			},
+
+			requested:  3,
+			lastPermit: lo.ToPtr[uint](2),
+			factor:     1,
+
+			// FIXME: This is actually incorrect / not working. The request should be rejected with
+			// only 2 reserved, because otherwise we end up over the total (as seen below).
+			verdict: "",
+			podAfter: pod{
+				reserved: 3,
+				buffer:   0,
+				pressure: 0,
+			},
+			nodeAfter: node{
+				reserved: 13,
+				buffer:   0,
+				pressure: 0,
+			},
+		},
+		{
+			name:   "DownscaleApprovedAboveLimits",
+			podMin: 1,
+			podMax: 5,
+			podBefore: pod{
+				reserved: 4,
+				buffer:   0,
+				pressure: 0,
+			},
+			nodeTotal: 10,
+			nodeBefore: node{
+				reserved: 15,
+				buffer:   0,
+				pressure: 0,
+			},
+
+			requested:  2,
+			lastPermit: lo.ToPtr[uint](4),
+			factor:     1,
+
+			verdict: "Register 4 -> 2 (pressure 0 -> 0); node reserved 15 -> 13 (of 10), node capacityPressure 0 -> 0 (0 -> 0 spoken for)",
+			podAfter: pod{
+				reserved: 2,
+				buffer:   0,
+				pressure: 0,
+			},
+			nodeAfter: node{
+				reserved: 13,
+				buffer:   0,
+				pressure: 0,
+			},
+		},
+		{
+			name:   "UpscaleApprovedUpToPermitAboveLimits",
+			podMin: 1,
+			podMax: 4,
+			podBefore: pod{
+				reserved: 2,
+				buffer:   0,
+				pressure: 0,
+			},
+			nodeTotal: 10,
+			nodeBefore: node{
+				reserved: 12,
+				buffer:   0,
+				pressure: 0,
+			},
+
+			requested:  3,
+			lastPermit: lo.ToPtr[uint](3),
+			factor:     1,
+
+			verdict: "Register 2 -> 3 (pressure 0 -> 0); node reserved 12 -> 13 (of 10), node capacityPressure 0 -> 0 (0 -> 0 spoken for)",
+			podAfter: pod{
+				reserved: 3,
+				buffer:   0,
+				pressure: 0,
+			},
+			nodeAfter: node{
+				reserved: 13,
+				buffer:   0,
+				pressure: 0,
+			},
+		},
+	}
```

Comment:
Probably related to the FIXMEs above, found something in the `trans.go`:

First we set `forceApprovalMinimum` to:
https://github.com/neondatabase/autoscaling/blob/2bf8d3e9e703e902f5963252d46da222528fa1ef/pkg/plugin/trans.go#L191-L194

Then we use it as the minimum increase here:
https://github.com/neondatabase/autoscaling/blob/2bf8d3e9e703e902f5963252d46da222528fa1ef/pkg/plugin/trans.go#L315-L316

Currently it means that that plugin always allows scaling up to 2 times from the `lastPermit` value, probably a bug?

Reply:
Exactly, yup (previously opened a fix PR in #1025).